maven-plugin/src/main/java/org/openmrs/maven/plugins/Build.java

```diff
@@ -160,13 +187,113 @@ private boolean buildCoreIfWatched(Server server) throws MojoFailureException {
+                    installLocalNodeAndNpm(projectNodeAndNpmVersion.get(NODE_VERSION_KEY), projectNodeAndNpmVersion.get(NPM_VERSION_KEY));
+                }
+                else {
+                    wizard.showMessage("Using default npm and nodejs versions");
+                    installLocalNodeAndNpm(SDKConstants.NODE_VERSION, SDKConstants.NPM_VERSION);
+                }
+            }
+            else {
+                isUsingSystemNpmAndNodejs = true;
+                wizard.showMessage("Using system npm and nodejs");
+                wizard.showMessage("-npm: " + systemNodeAndNpmVersion.get(NPM_VERSION_KEY)
+                       + "\n" + "-nodejs: " + systemNodeAndNpmVersion.get(NODE_VERSION_KEY) + "\n");
+            }
+        }
+
+        installNodeModules(isUsingSystemNpmAndNodejs);
+        runOwaBuild(isUsingSystemNpmAndNodejs);
+
+        wizard.showMessage("Build done.");
+    }
+
+    private boolean checkIfVersionsAreDefined(Map <String, String> versions) {
+        return StringUtils.isNotBlank(versions.get(NPM_VERSION_KEY)) && StringUtils.isNotBlank(versions.get(NODE_VERSION_KEY));
+    }
+
+    protected Map<String, String> getBatchNodeAndNpmVersion() {
+        if (nodeVersion != null && !nodeVersion.startsWith("v")) {
+            nodeVersion = "v" + nodeVersion;
+        }
+        Map<String, String> result = new HashMap<>();
+        result.put(NPM_VERSION_KEY, npmVersion);
+        result.put(NODE_VERSION_KEY, nodeVersion);
+        return result;
+    }
+
+    protected Map<String, String> getSystemNodeAndNpmVersion() {
+        Map<String, String> result = new HashMap<>();
+        result.put(NPM_VERSION_KEY, getSystemNpmVersion());
+        result.put(NODE_VERSION_KEY, getSystemNodeVersion());
+        return result;
+    }
+
+    protected Map<String, String> getProjectNpmAndNodeVersionFromPackageJson() {
+        Map<String, String> result = new HashMap<>();
+        result.put(NPM_VERSION_KEY, getProjectNpmVersionFromPackageJson());
+        result.put(NODE_VERSION_KEY, getProjectNodeVersionFromPackageJson());
+        return result;
+    }
+
+    private String getSystemNpmVersion() {
+        return runProcessAndGetFirstResponseLine("npm", "-v");
+    }
+    private String getSystemNodeVersion() {
+        return runProcessAndGetFirstResponseLine("node", "-v");
+    }
+
+    private String getProjectNpmVersionFromPackageJson() {
+        return getPropertyValueFromPropertiesJsonFile(PACKAGE_JSON_FILENAME, NPM_VERSION_KEY);
+    }
+    private String getProjectNodeVersionFromPackageJson() {
+        return getPropertyValueFromPropertiesJsonFile(PACKAGE_JSON_FILENAME, NODE_VERSION_KEY);
+    }
+
+    protected static String getPropertyValueFromPropertiesJsonFile(String jsonFilename, String key) {
+        JSONParser parser = new JSONParser();
+        Object obj;
+        try {
+            obj = parser.parse(new FileReader(jsonFilename));
```

Comment:
Not sure about the library you used, but typically you have to close a file after reading....